api/v1beta1/openstackdataplaneservice_types.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+	ansibleeev1alpha1 "github.com/openstack-k8s-operators/openstack-ansibleee-operator/api/v1alpha1"
+)
+
+// OpenStackDataPlaneServiceSpec defines the desired state of OpenStackDataPlaneService
+type OpenStackDataPlaneServiceSpec struct {
+	// Label to use for service
+	// +kubebuilder:validation:Optional
+	Label string `json:"label,omitempty"`
+
+	// Role is the description of an Ansible Role
+	// +kubebuilder:validation:Optional
+	Role *ansibleeev1alpha1.Role `json:"role,omitempty"`
+}
```

Comment:
@karelyatin @slawqo I think this discussion are interesting to you too as it involves the design of deploying networking services on edpm node.

Reply:
The OpenStackDataPlaneService CRD is ansible specific in that it will only equate to OpenStackAnsibleEE CR's. The ansible roles or plays executed by the OpenStackAnsibleEE will be fully configurable with the existing ansibleVars interface we already have.

For things like RabbitMQCluster/TransportURL, the dataplane-operator can look up the associated values given an OpenStackControlPlane name, and set those values within ansibleVars for any ansible role to consume.

I'm not aware of plans by the networking folks to change the way that the networking related services are currently being deployed between dataplane-operator/edpm-ansible. In the EDPM sync meeting, we brought this up and I believe @slawqo was there and indicated there was no planned work to change that. If that is incorrect, we need to draw attention to that, in terms of what issues exist and who is planning on doing the work within the other operators and dataplane-operator.

Reply:
> For things like RabbitMQCluster/TransportURL, the dataplane-operator can look up the associated values given an OpenStackControlPlane name, and set those values within ansibleVars for any ansible role to consume.

Unfortunately this introduce a dependency cycle. Similar to what we had when TransportURL CRD was defined by OpenstackOperator. We later moved that CRD to a separate infra-operator to break the cycle.

Any change that result in a new version of openstack-operator will cause a new version of the apis module of the openstack-operator the dataplane-operator depends on. So the dataplane-operator will bump the version of the openstack-operator apis module in its go.mod file (and do the necessary adaptation work if needed). This will result in a new dataplane-operator version (i.e. new git hash on master). This in return will trigger a version bump in the openstack-operator's go.mod file for the dataplane-operator dependency. When that bump is merged in openstack-operator a new version of the opeanstack-operator apis module (a new git hash) is produced and the cycle restarts.

With the nova-operator approach we don't have a dependency cycle as only dataplane-operator depends on nova-operator's api module but nova-operator does not mention dataplane-operator in any of its go.mod files.

> I'm not aware of plans by the networking folks to change the way that the networking related services are currently being deployed between dataplane-operator/edpm-ansible. In the EDPM sync meeting, we brought this up and I believe @slawqo was there and indicated there was no planned work to change that. If that is incorrect, we need to draw attention to that, in terms of what issues exist and who is planning on doing the work within the other operators and dataplane-operator.

I agree we need to settle this sooner than later.

Reply:
> > For things like RabbitMQCluster/TransportURL, the dataplane-operator can look up the associated values given an OpenStackControlPlane name, and set those values within ansibleVars for any ansible role to consume.
> 
> Unfortunately this introduces a dependency cycle. Similar to what we had when TransportURL CRD was defined by OpenstackOperator. We later moved that CRD to a separate infra-operator to break the cycle.

This patch doesn't introduce that kind of cycle though. I don't see that kind of import in any of this code.

I just tested this patch in my environment and `oc edit openstackdataplanes.dataplane.openstack.org` reveals that the `edpm_ovn_metadata_agent_DEFAULT_transport_url` ansible var is set the same way it has been set.

Reply:
It's my understanding that these issues have been resolved in offline conversations. Please chime in if there's still any objections to merging this.

Reply:
To clarify, the discussion about changing ovn deployment from dataplane-operator to ovn-operator seemed to settle around the consensus that what we have now with dataplane-operator is fine to proceed. If the OVN folks want to look at migrating that out of dataplane-operator later, then that is still possible.

Reply:
I rest my case here. As @fultonj stated above the current patch does not introduce any dependency cycle. My point about the cycle was a response to @slagle proposal to look up the RabbitMQCluster/TransportURL CR from dataplane-operator.

Reply:
I do feel that the current way we fill transport URL configuration via install_yamls is not a nice end user interface though. But that discussion is definitely independent from this PR.